gapis/api/transform2/transform_chain.go

```diff
@@ -0,0 +1,190 @@
+		if err != nil {
+			log.W(ctx, "Begin Transform Error [%v] : %v", transform, err)
+			return err
+		}
+
+		if transform.RequiresAccurateState() {
+			if err = mutateAndWrite(ctx, 0, cmds, chain.out); err != nil {
+				return err
+			}
+
+			transform.ClearTransformResources(ctx)
+		}
+	}
+
+	if err = mutateAndWrite(ctx, 0, cmds, chain.out); err != nil {
+		return err
+	}
+
+	for _, transform := range chain.transforms {
+		transform.ClearTransformResources(ctx)
+	}
+
+	return nil
+}
+
+func (chain *transformChain) endChain(ctx context.Context) error {
+	var err error
+	cmds := make([]api.Cmd, 0)
+
+	for _, transform := range chain.transforms {
+		cmds, err = transform.EndTransform(ctx, cmds, chain.out.State())
+		if err != nil {
+			log.W(ctx, "End Transform Error [%v] : %v", transform, err)
+			return err
+		}
+
+		if transform.RequiresAccurateState() {
+			if err = mutateAndWrite(ctx, 0, cmds, chain.out); err != nil {
+				return err
+			}
+
+			transform.ClearTransformResources(ctx)
+		}
+	}
+
+	if err = mutateAndWrite(ctx, 0, cmds, chain.out); err != nil {
+		return err
+	}
+
+	for _, transform := range chain.transforms {
+		transform.ClearTransformResources(ctx)
+	}
+
+	return nil
+}
+
+func (chain *transformChain) transformCommand(ctx context.Context, inputCmd api.Cmd, id api.CmdID) error {
+	var err error
+	cmds := []api.Cmd{inputCmd}
+
+	for _, transform := range chain.transforms {
+		cmds, err = transform.TransformCommand(ctx, id, cmds, chain.out.State())
+		if err != nil {
+			log.W(ctx, "Error on Transform on cmd [%v:%v] with transform [%v] : %v", id, inputCmd, transform, err)
+			return err
+		}
+
+		if transform.RequiresAccurateState() {
+			if err = mutateAndWrite(ctx, id, cmds, chain.out); err != nil {
+				return err
+			}
+
+			transform.ClearTransformResources(ctx)
+		}
+	}
+
+	if err = mutateAndWrite(ctx, id, cmds, chain.out); err != nil {
+		return err
+	}
+
+	for _, transform := range chain.transforms {
+		transform.ClearTransformResources(ctx)
+	}
+
+	return nil
+}
+
+func (chain *transformChain) IsEndOfCommands() bool {
+	return chain.hasEnded
+}
+
+func (chain *transformChain) GetNextTransformedCommands(ctx context.Context) error {
+	if !chain.hasBegun {
+		chain.hasBegun = true
+		return chain.beginChain(ctx)
+	}
+
+	if chain.generator.IsEndOfCommands() {
+		if !chain.hasEnded {
+			chain.hasEnded = true
+			return chain.endChain(ctx)
+		}
+
+		return nil
+	}
+
+	currentCommand := chain.generator.GetNextCommand(ctx)
+	if config.DebugReplay {
+		log.I(ctx, "Transforming... (%v:%v)", chain.currentCommandID, currentCommand)
+	}
+
+	err := chain.transformCommand(ctx, currentCommand, chain.currentCommandID)
+	if err != nil {
+		log.E(ctx, "Replay error (%v:%v): %v", chain.currentCommandID, currentCommand, err)
+	}
+	chain.currentCommandID++
+	return err
+}
+
+func handleInitialState(state *api.GlobalState) (*api.GlobalState, error) {
+	// Melih TODO: Currently this is not in use as we do not need the feature yet
```

Comment:
You could check to see if anyone is saying needsAccurateState == true and panic with a todo message.

Reply:
Since you haven't implemented this yet, it feels safer than silently continuing!